### PR TITLE
Calendar integration

### LINKED ISSUE #215: Calendar Integration
# Description
Fully integrate the Calendar with the backend

# Acceptance Criteria
- [ ] Integration is complete
- [ ] Unit tests are updated
- [ ] Unit tests pass

### PR DESCRIPTION
**Pull request by elijah-parrish**
closes #215

### CODE REVIEW

File: database/mock_data/PopulateDB.sql [hunk 1]
```diff
@@ -145,14 +145,13 @@ insert into member (member_type_id, first_name, last_name, phone_number, occupat
...
-insert into member (uuid, member_type_id, first_name, last_name, phone_number, occupation, email, birthdate, date_joined, active) values ('28m', 3, 'Vasilis', 'Mertel', '561-878-3783', 'Developer II', 'vmertel2o@parallels.com', '1976-11-30', '2019-02-03', 1);
 insert into member (member_type_id, first_name, last_name, phone_number, occupation, email, birthdate, date_joined, active) values (3, 'Kathie', 'O''Scannill', '760-880-8632', 'Tax Accountant', 'koscannill2p@myspace.com', '2009-12-18', '2020-10-15', 1);
 insert into member (member_type_id, first_name, last_name, phone_number, occupation, email, birthdate, date_joined, active) values (2, 'Patin', 'Lelliott', '913-810-0953', 'Senior Financial Analyst', 'plelliott2q@multiply.com', '2005-12-11', '2018-05-19', 1);
+insert into member (uuid, member_type_id, first_name, last_name, phone_number, occupation, email, birthdate, date_joined, active) values ('28m', 3, 'Vasilis', 'Mertel', '561-878-3783', 'Developer II', 'vmertel2o@parallels.com', '1976-11-30', '2019-02-03', 1);
```
Comment (matedro) [inline database/mock_data/PopulateDB.sql:150]: I think this is the reason why the integration test is failing. Why did you do this?

File: database/mock_data/PopulateDB_Part2.sql [hunk 2]
```diff
@@ -29,4 +29,7 @@ update membership set membership_admin_id = 1 where membership_id = 2; -- this w
 update member m
 left join membership ms on m.membership_id = ms.membership_id
 set membership_admin_id = m.member_id
-where m.member_type_id = 2;
\ No newline at end of file
+where m.member_type_id = 2;
+
+CALL sp_event_job_generation('2022-04-01T08:00:00', '2022-04-01T16:00:00', 1, '2022 Last Race', 'test empty event', @ignore);
```
Comment (matedro) [inline database/mock_data/PopulateDB_Part2.sql:34]: Please put this up top with the rest of the calls to `sp_event_job_generation`

Comment (elijah-parrish) [reply database/mock_data/PopulateDB_Part2.sql:34]: This was intentional (per Jacob's idea) to create an event with jobs which aren't assigned to members yet

File: database/mock_data/PopulateDB.sql (hunk 1 above)
Comment (elijah-parrish) [reply database/mock_data/PopulateDB.sql]: Honestly... no idea

File: database/mock_data/PopulateDB_Part2.sql (hunk 2 above)
Comment (matedro) [reply database/mock_data/PopulateDB_Part2.sql:34]: No, I was just trying to keep it organized. If there's a reason for it then this is fine 👍

File: database/mock_data/PopulateDB_Part2.sql [hunk 3]
```diff
@@ -1,14 +1,14 @@
 ### GEN THE JOBS
-CALL sp_event_job_generation('2020-02-01T08:00:00', '2020-02-02T16:00:00', 1, 'The First Race', 'test first race!', @ignore);
```
Comment (matedro) [inline database/mock_data/PopulateDB_Part2.sql:2]: If you change these back to 2020, the integration tests should pass. Looks like this got lost in the merge.

File: src/__tests__/integration/workPoints.test.ts [hunk 4]
```diff
@@ -68,7 +68,7 @@ describe('GET /workPoints/byMember/:id', () => {
 
     it('Returns current-year work points for member', async () => {
         const memberId = 1;
-        const expWorkPoints = { total: 3 };
+        const expWorkPoints = { total: 8 };
```
Comment (matedro) [inline src/__tests__/integration/workPoints.test.ts:71]: Ah, I think you'll also have to revert the changes in this file.

File: frontend/src/components/EventCalendar.tsx [hunk 5]
```diff
@@ -42,65 +36,167 @@ const customMessages: Messages = {
     agenda: 'Agenda',
 };
 
-async function getSelectedJobAttendees(): Promise<any> {
-    const attendees = await getJobAttendees();
-    return attendees;
-}
+async function getCalendarEventsLocal(token: string) {
```
Comment (cjs8487) [inline frontend/src/components/EventCalendar.tsx:39]: don't call it local

File: frontend/src/components/EventCalendar.tsx [hunk 6]
```diff
@@ -42,65 +36,167 @@ const customMessages: Messages = {
...
     const { onClose: onViewEventClose, isOpen: isViewEventOpen, onOpen: onViewEventOpen } = useDisclosure();
     const { onClose: onSignUpClose, isOpen: isSignUpOpen, onOpen: onSignUpOpen } = useDisclosure();
-    const [selectedEvent, setSelectedEvent] = useState<any>();
-    const [eventAttendees, setAttendees] = useState<any>();
+    const [selectedEvent, setSelectedEvent] = useState<Event | Job>();
     const [familyMembers, setFamilyMembers] = useState<any>();
+    const [calendarEvents, setCalendarEvents] = useState<Array<Job | Event>>([]);
+
+    function isEvent(event: Event | Job): event is Event {
+        if ((event as Event).eventType) {
```
Comment (cjs8487) [inline frontend/src/components/EventCalendar.tsx:60]: use `'eventType' in event`

Comment (cjs8487) [reply frontend/src/components/EventCalendar.tsx:60]: this could also be moved outside the component as it does not depend on anything within

File: frontend/src/components/EventCalendar.tsx [hunk 7]
```diff
@@ -42,65 +36,167 @@ const customMessages: Messages = {
...
+            return true;
+        }
+        // else, its a Job
+        return false;
+    }
+
+    function deleteEventWasSuccessful(calendarEvent: DeletedEvent | ErrorResponse): calendarEvent is DeletedEvent {
```
Comment (cjs8487) [inline frontend/src/components/EventCalendar.tsx:67]: there are simpler checks for this by looking for the reason field on the response - no need to make it complicated

File: frontend/src/components/EventCalendar.tsx [hunk 8]
```diff
@@ -42,65 +36,167 @@ const customMessages: Messages = {
...
+        if ((calendarEvent as DeletedEvent).eventId) {
+            return true;
+        }
+        // else, its an error
+        return false;
+    }
+
+    function deleteJobWasSuccessfull(calendarEvent: DeletedJob | ErrorResponse): calendarEvent is DeletedJob {
```
Comment (cjs8487) [inline frontend/src/components/EventCalendar.tsx:75]: same here

File: frontend/src/components/EventCalendar.tsx [hunk 9]
```diff
@@ -42,65 +36,167 @@ const customMessages: Messages = {
...
+        if ((calendarEvent as DeletedJob).jobId) {
+            return true;
+        }
+        // else, its an error
+        return false;
+    }
+
+    async function createEventLocal(newEvent: PostNewEventRequest) {
```
Comment (cjs8487) [inline frontend/src/components/EventCalendar.tsx:83]: no "local" functions - also moment is deprecated

File: frontend/src/components/EventCalendar.tsx [hunk 10]
```diff
@@ -42,65 +36,167 @@ const customMessages: Messages = {
...
+                const newCalendarEvents = calendarEvents.filter((e: any) => e.jobId !== selectedEvent?.jobId);
+                setCalendarEvents(newCalendarEvents);
+            } else {
+                // eslint-disable-next-line no-console
+                console.log(response.reason);
+            }
+        }
+    }
+
+    async function getFamilyMembersLocal() {
```
Comment (cjs8487) [inline frontend/src/components/EventCalendar.tsx:127]: local

File: frontend/src/components/EventCalendar.tsx [hunk 11]
```diff
@@ -42,65 +36,167 @@ const customMessages: Messages = {
...
+        if (state.user) {
```
Comment (cjs8487) [inline frontend/src/components/EventCalendar.tsx:128]: you can use `state.user?.token` to assert that it exists in place

File: frontend/src/components/EventCalendar.tsx [hunk 12]
```diff
@@ -42,65 +36,167 @@ const customMessages: Messages = {
...
+            const res: GetMemberListResponse = await getFamilyMembers(state.token, state.user.membershipId);
+            if ('reason' in res) {
+                console.log(res.reason);
```
Comment (cjs8487) [inline frontend/src/components/EventCalendar.tsx:131]: console log

File: frontend/src/components/EventCalendar.tsx [hunk 13]
```diff
@@ -141,24 +237,29 @@ export default function EventCalendar(props: EventCalendarProps) {
...
-                selectedEvent && eventAttendees && (
+                selectedEvent && (
                     <SelectedEventModal
                         isOpen={isViewEventOpen}
                         onClose={onViewEventClose}
                         selectedEvent={selectedEvent}
                         onSignUpOpen={onSignUpOpen}
-                        // attendeesList={eventAttendees}
                         admin={state.user?.memberType === 'Admin'}
+                        // eslint-disable-next-line react/jsx-no-bind
```
Comment (cjs8487) [inline frontend/src/components/EventCalendar.tsx:247]: are these still needed?

File: frontend/src/components/FamilySignUpModal.tsx [hunk 14]
```diff
@@ -0,0 +1,107 @@
...
+import { Member } from '../../../src/typedefs/member';
+import { Job, PatchJobRequest } from '../../../src/typedefs/job';
+import { UserContext } from '../contexts/UserContext';
+
+interface modalProps {
+  isOpen: boolean,
+  onClose: () => void,
+  familyMembers: Member[],
+  selectedEvent: any, // TODO: bugfix needed, Event | Job breaks here
+  // eslint-disable-next-line no-unused-vars
```
Comment (cjs8487) [inline frontend/src/components/FamilySignUpModal.tsx:25]: disable bad

File: frontend/src/components/FamilySignUpModal.tsx [hunk 15]
```diff
@@ -0,0 +1,107 @@
...
+import moment from 'moment';
+import { Member } from '../../../src/typedefs/member';
+import { Job, PatchJobRequest } from '../../../src/typedefs/job';
+import { UserContext } from '../contexts/UserContext';
+
+interface modalProps {
+  isOpen: boolean,
+  onClose: () => void,
+  familyMembers: Member[],
+  selectedEvent: any, // TODO: bugfix needed, Event | Job breaks here
```
Comment (cjs8487) [inline frontend/src/components/FamilySignUpModal.tsx:24]: is this bug fix still needed

File: frontend/src/components/FamilySignUpModal.tsx [hunk 16]
```diff
@@ -0,0 +1,107 @@
...
+
+    function isJob(selectedEvent: Event | Job): selectedEvent is Job {
+        if ((selectedEvent as Job).jobId) {
+            return true;
+        }
+        // else, its an Event
+        return false;
+    }
+
+    async function generateJobSignUpPatch(memberId: number) {
```
Comment (cjs8487) [inline frontend/src/components/FamilySignUpModal.tsx:40]: if this does not depend on local state, move it outside the function. If it does. wrap it in `useCallback(() => ... ,[states])`

File: frontend/src/components/SelectedEventModal.tsx [hunk 17]
```diff
@@ -15,39 +16,55 @@ import {
     HStack,
     VStack,
     Text,
-    // ListItem,
-    // UnorderedList,
 } from '@chakra-ui/react';
+import moment from 'moment';
```
Comment (cjs8487) [inline frontend/src/components/SelectedEventModal.tsx:20]: moment is still deprecated

File: frontend/src/components/SelectedEventModal.tsx [hunk 18]
```diff
@@ -15,39 +16,55 @@ import {
...
 interface modalProps {
   isOpen: boolean,
   onClose: () => void,
   selectedEvent: any,
   onSignUpOpen: () => void;
-  // attendeesList: any[], // TODO this should match our typing
-  admin: boolean
+  admin: boolean;
+  deleteEvent: () => void;
+  // eslint-disable-next-line no-unused-vars
```
Comment (cjs8487) [inline frontend/src/components/SelectedEventModal.tsx:33]: bad disable

File: frontend/src/components/SelectedEventModal.tsx [hunk 19]
```diff
@@ -15,39 +16,55 @@ import {
...
-    // if (event.type === 'work') {
-    //     await deleteJob('TestingToken', event);
-    // } else {
-    //     await deleteEvent('TestingToken', event);
-    // }
-    // eslint-disable-next-line no-console
-    console.log(event);
-    // Update state?
-}
+    function isJob(selectedEvent: Event | Job): selectedEvent is Job {
```
Comment (cjs8487) [inline frontend/src/components/SelectedEventModal.tsx:40]: again there are easier ways to do this, it is also duplicated logic as is

File: frontend/src/components/SelectedEventModal.tsx [hunk 20]
```diff
@@ -15,39 +16,55 @@ import {
...
+        if ((selectedEvent as Job).jobId) {
+            return true;
+        }
+        // else, its an Event
+        return false;
+    }
+
+    async function generateJobSignUpPatch() {
```
Comment (cjs8487) [inline frontend/src/components/SelectedEventModal.tsx:48]: this looks to be duplicated logic as well

File: frontend/src/controller/job.ts [hunk 21]
```diff
@@ -8,12 +8,18 @@ import {
     PostCloneJobResponse,
     PostNewJobRequest,
     PostNewJobResponse,
+    Job,
 } from '../../../src/typedefs/job';
+import { ErrorResponse } from '../../../src/typedefs/errorResponse';
+
+function isJobList(res: Job[] | ErrorResponse): res is Job[] {
```
Comment (cjs8487) [inline frontend/src/controller/job.ts:15]: again simpler logic

File: frontend/src/controller/job.ts [hunk 22]
```diff
@@ -24,20 +30,34 @@ export async function getJobList(token: string, queryType?: string, filterType?:
...
     const response = await fetch(`${process.env.REACT_APP_API_URL}/api/job/list`, {
         method: 'GET',
-        mode: 'no-cors',
+        mode: 'cors',
         headers: generateHeaders(token),
     });
     return response.json();
 }
 
+export async function getCalendarJobs(token: string) {
```
Comment (cjs8487) [inline frontend/src/controller/job.ts:47]: this logic also looks duplicated

File: frontend/src/pages/CalendarPage.tsx [hunk 23]
```diff
@@ -1,37 +1,52 @@
...
+import { getCalendarEvents } from '../controller/event';
+import { getCalendarJobs } from '../controller/job';
+import { Event } from '../../../src/typedefs/event';
+import { Job } from '../../../src/typedefs/job';
 
-// TODO: should we also be getting all jobs and displaying them as well?
-async function getEventListLocal(): Promise<any> {
-    const events = await getCalendarEventList();
-    return events;
+async function getCalendarEventsLocal(token: string) {
```
Comment (cjs8487) [inline frontend/src/pages/CalendarPage.tsx:13]: local

File: frontend/src/pages/Dashboard.tsx [hunk 24]
```diff
@@ -18,6 +18,7 @@ async function getEventCardPropsLocal(token: string): Promise<any | undefined> {
     return props;
 }
 
+// eslint-disable-next-line no-unused-vars
```
Comment (cjs8487) [inline frontend/src/pages/Dashboard.tsx:21]: bad disable

**Review by cjs8487 — CHANGES_REQUESTED** [review]

File: frontend/src/components/CreateEventModal.tsx [hunk 25]
```diff
@@ -17,40 +17,53 @@ import {
...
+        if (getEventTypeListWasSuccessful(res)) {
+            setEventTypes(res);
+        }
+    }
 
     useEffect(() => {
         async function getData() {
-            const types = await getMockedEventTypeList();
-            setEventTypes(types);
+            await getEventTypeListLocal(state.token);
```
[the remark this answers is not in the document]
Comment (elijah-parrish) [reply frontend/src/components/CreateEventModal.tsx:66]: I'm not sure I understand, is this not in the effect?

[the remark this answers is not in the document]
Comment (cjs8487) [reply frontend/src/components/CreateEventModal.tsx:66]: You're calling a function that is external to the effect, which is the problem - effect callbacks are optimized in part based on their dependency array, but functions that they call out too are not and can still potentially be referencing old values because of the way function declarations work in function components

File: frontend/src/components/SelectedEventModal.tsx (hunk 20 above)
Comment (elijah-parrish) [reply frontend/src/components/SelectedEventModal.tsx:48]: What are your thoughts on putting this in the job controller? Utils?

**Review by cjs8487 — APPROVED** [review]
